### PR TITLE
[QMS 396] Save route sub-points as ordinary route points in case of track routing

### PR DESCRIPTION
**Pull request by UEckhardt**
### What is the linked issue for this pull request:
[comment]: # (Add the ticket number behind QMS-# )

QMS-#396

### What you have done:
[comment]: # (Describe roughly.)
Implement the basic track to route conversion via a new icon in the track context menu and also to the track range menu so that a route can be created from a track range.

Add an option to the Workspace setup dialog to configure if the track to route conversion will include the sub-points of the track as route points.

German translation is added.

### Steps to perform a simple smoke test:
[comment]: # (List the steps.)

1. Create a small track with 5 points and perform an auto-routing for this track,
2. Open the setup workspace dialog and ensure that the new option "Include sub-points.." is unchecked.
3. Select the track and click on the new route icon in the track context menu.
4. A new route without sub-points is now created.
5. Select the track, click on the range of points button and select a subset of the track.
6. Click on the new route icon in the select range menu.
7. A new route without sub-points is created only for the subset of this track .
8. Open the setup workspace dialog and select the new option "Include sub-points..".
9. Select the track and click on the new route icon in the track context menu.
10. A new route including sub-points is now created.
11. Select the track, click on the range of points button and select a subset of the track.
12. Click on the new route icon in the select range menu.
13. A new route including sub-points is created only for the subset of this track .

### Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):

- [X] yes

### Is every user facing string in a tr() macro?

- [X] yes

### Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.

- [X] yes, I didn't forget to change changelog.txt

### CODE REVIEW

**Comment by kiozen** [comment]
Thanks for the PR. I am currently travelling and will take care of it as soon as possible,

File: src/qmapshack/gis/db/CSetupWorkspace.cpp [hunk 1]
```diff
@@ -36,6 +36,7 @@ CSetupWorkspace::CSetupWorkspace(CGisWorkspace* workspace, QWidget* parent)
     checkDbUpdate->setChecked(cfg.value("listenUpdate", false).toBool());
     linePort->setText(cfg.value("port", "34123").toString());
     checkDeviceSupport->setChecked(cfg.value("device support", true).toBool());
+    checkRouteConversion->setChecked(cfg.value("saveSubPoints", false).toBool());
```
Comment (kiozen) [inline src/qmapshack/gis/db/CSetupWorkspace.cpp:39]: This is not the right place nor the right way. Track to route conversion has nothing to do with the workspace. Please use dialog to ask for that option. Within that dialog the user can change the name, too. And you could add some point limit and apply the Douglas Peuker algorithm to reduce points.

File: src/qmapshack/gis/trk/CTrackData.cpp [hunk 2]
```diff
@@ -140,6 +140,39 @@ void CTrackData::getPolyline(SGisLine& l) const
     }
 }
 
+void CTrackData::getPolylineRange(SGisLine& l, qint32 rangeStart, qint32 rangeEnd, bool getSubPixel) const
+{
+    l.clear();
+    for (const trkseg_t &cur_seg : segs)
+    {
+        for (const trkpt_t &pt : cur_seg.pts)
```
Comment (kiozen) [inline src/qmapshack/gis/trk/CTrackData.cpp:148]: Naming convention. no snake_case. camelCase.  In that case `seg` is totally fine.

File: src/qmapshack/gis/trk/CTrackData.cpp [hunk 3]
```diff
@@ -140,6 +140,39 @@ void CTrackData::getPolyline(SGisLine& l) const
...
+        {
+            if((pt.idxTotal < rangeStart) && (rangeStart != NOIDX))
+            {
+                continue;
+            }
+            if((pt.idxTotal > rangeEnd) && (rangeEnd != NOIDX))
+            {
+                break;
+            }
+            if(!pt.isHidden())
```
Comment (kiozen) [inline src/qmapshack/gis/trk/CTrackData.cpp:158]:
```
if(pt.isHidden())
{
    continue;
}
```

Is easier to read

File: src/qmapshack/gis/trk/CTrackData.h [hunk 4]
```diff
@@ -244,6 +244,20 @@ class CTrackData
...
+
+           @param l Converted Polyline.
+           @param rangeStart Start index of the conversion. If set to NOIDX, conversion starts at
+                  the beginning of the track.
+           @param rangeEnd End index of the conversion. If set to NOIDX, conversion end at
+                  the last point of the track.
+           @param getSubPixel If getSubPixel is true all sub-pixels are converted to a point.
+                  If set to false the sub-pixels are ignored.
+    */
+    void getPolylineRange(SGisLine& l, qint32 rangeStart, qint32 rangeEnd, bool getSubPixel = false) const;
```
Comment (kiozen) [inline src/qmapshack/gis/trk/CTrackData.h:260]: Well documented! Thanks! :+1:

**Review by kiozen — CHANGES_REQUESTED** [review]
Please revert all `*.ts` files. This will be a pain in the ass to merge if there are other changes in dev. (and you have to fix the conflicts ;) ). That's why we decided to do translations always after the commit and that it's ok to do it without PR.

The thing with the options has to be solved. I know you are no fan of those dialogs. Me neither. But at this point it makes sens. Especially if you want to change the name as it can be done for route to track conversion. And it's a good way to set other/future options to simplify/control the conversion.

**Comment by kiozen** [comment]
There is another point that needs to be done: If you click in the workspace tree on an item withe the right mouse you get a context menu with the same actions as in the on-screen toolbox. You have to add your feature, too.

**Comment by kiozen** [comment]
Merging the dev branch into you branch is a bad idea. Please rebase on dev. By that your commits stay in a linear order and we do not end up in trouble when merging into dev.

File: src/qmapshack/gis/db/CSetupWorkspace.cpp (hunk 1 above)
Comment (UEckhardt) [reply src/qmapshack/gis/db/CSetupWorkspace.cpp]: I have pushed a new version with a new track to route dialog. This dialog combines setting/editing the route name, selecting a new project and sub-point conversion.

I will have a look at the Douglas Peuker algorithm and see if I could incorporate this algorithm also in the dialog.

**Comment by UEckhardt** [comment]
Due to the rebase I have garbled my repository :-( , so I had to setup the fork again. I will close this pull request and resent a new one.

**Pull request closed without merging**